hs_discover/views.py

```diff
@@ -1,57 +1,224 @@
+            filters = json.loads(qs.get('filter'))
+            # filter values expect lists, for example discoverapi/?filter={"owner":["Firstname Lastname"]}
+            if filters.get('author'):
+                sqs = sqs.filter(author__in=filters['author'])
+            if filters.get('owner'):
+                sqs = sqs.filter(owner__in=filters['owner'])
+            if filters.get('subject'):
+                sqs = sqs.filter(subject__in=filters['subject'])
+            if filters.get('contributor'):
+                sqs = sqs.filter(contributor__in=filters['contributor'])
+            if filters.get('type'):
+                sqs = sqs.filter(resource_type__in=list(filters['type']))
+            if filters.get('availability'):
+                sqs = sqs.filter(availability__in=filters['availability'])
+            if filters.get('geofilter'):
+                sqs = sqs.filter(north__range=[-90, 90])
+            if filters.get('date'):
+                try:
+                    datefilter = DateRange(start=datetime.datetime.strptime(filters['date'][0], '%Y-%m-%d'),
+                                           end=datetime.datetime.strptime(filters['date'][1], '%Y-%m-%d'))
 
-        vocab = []
-        for result in sqs:
-            if result.title:
-                vocab.extend(result.title.split(' '))
-            if result.subject:
-                vocab.extend(result.subject)
+                    # (datefilter.start < start_date < datefilter.end) or (start_date < datefilter.start)
+                    sqs = sqs.exclude(start_date__gt=datefilter.end).exclude(end_date__lt=datefilter.start)
+                except ValueError as e:
+                    pass  # ignore bad values and don't filter which is the correct action
+        except TypeError as type_ex:
+            pass  # no filters passed "the JSON object must be str, bytes or bytearray not NoneType"
 
-        vocab = [x for x in vocab if len(x) > 2]
-        vocab = list(set(vocab))
-        vocab = sorted(vocab)
+        except json.JSONDecodeError as parse_ex:
+            return JsonResponse({'message': 'Filter JSON parsing error - {}'.format(str(parse_ex)),
+                                 'received': request.query_params}, status=400)
 
-        if q:
-            sqs = sqs.filter(content=q).boost('keyword', 2.0)
+        except Exception as gen_ex:
+            logger.debug('hs_discover API - {}: {}'.format(type(gen_ex), str(gen_ex)))
+            return JsonResponse({'message': '{}'.format('{}: query error. Contact a server administrator.'
+                                                        .format(type(gen_ex)))}, status=520)
+
+        sqs = sqs.order_by(sort)
 
         resources = []
-        for result in sqs:
+
+        p = Paginator(sqs, self.perpage)
+
+        pnum = 1
+        if request.GET.get('pnum'):
+            pnum = request.GET.get('pnum')
+            pnum = max(1, int(pnum))
+            pnum = min(pnum, p.num_pages)
```

Comment:
if someone makes a request for a non-existing page number, I think it's better to respond with bad request rather than changing the page number to a valid one.

Reply:
This results in a negative User Experience or tons of coding in any client to open the response, look for the error, see if detail on the error was passed and verify that it was a page number error and not some other kind, then send another server request for a properly paged request. The use case is a user tapping the up or down arrow keys, or rapidly mouse clicking on the up/down value button in the input box. I highly recommend leaving this as-is.

Reply:
I'm not quite following and I don't think this is a heavy ask on the client side.  If it is a heavy ask on the client side... probably better to return an empty result for a non-existing page instead of magically changing the request to a valid page with zero feedback or logging.

Reply:
The intuition here is that searches return Page 1 with 0 results. I'll update the endpoint to return no results for invalid page numbers, such as negative numbers and zero.

Reply:
Commit bca65401f